### PR TITLE
feat: add client side retry to GeminiTextGenerator

### PR DESCRIPTION
**Pull request by GarrettWu**
Thank you for opening a Pull Request! Before submitting your PR, there are a few things you can do to make sure it goes smoothly:
- [ ] Make sure to open an issue as a [bug/issue](https://github.com/googleapis/python-bigquery-dataframes/issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [ ] Ensure the tests and linter pass
- [ ] Code coverage does not decrease (if any source code was changed)
- [ ] Appropriate docs were updated (if necessary)

b/385221339

### CODE REVIEW

File: bigframes/ml/llm.py [hunk 1]
```diff
@@ -945,6 +946,7 @@ def predict(
         top_k: int = 40,
         top_p: float = 1.0,
         ground_with_google_search: bool = False,
+        retry: int = 0,
```
Comment (jiaxunwu) [inline bigframes/ml/llm.py:949]: Can we rename it to max_retries?

File: bigframes/ml/llm.py [hunk 2]
```diff
@@ -1018,15 +1029,34 @@ def predict(
...
 
-        df = self._bqml_model.generate_text(X, options)
+        df_result = bpd.DataFrame(session=self._bqml_model.session)  # placeholder
+        df_fail = X
+        for _ in range(retry + 1):
+            df = self._bqml_model.generate_text(df_fail, options)
 
-        if (df[_ML_GENERATE_TEXT_STATUS] != "").any():
+            df_succ = df[df[_ML_GENERATE_TEXT_STATUS].str.len() == 0]
+            df_fail = df[df[_ML_GENERATE_TEXT_STATUS].str.len() > 0]
```
Comment (jiaxunwu) [inline bigframes/ml/llm.py:1038]: Can we just use df_succ to tell whether there is any failed row? Then it can reduce one filter op.

**Review by jiaxunwu — APPROVED** [review]

File: bigframes/ml/llm.py (hunk 1 above)
Comment (GarrettWu) [reply bigframes/ml/llm.py:949]: SG

File: bigframes/ml/llm.py (hunk 2 above)
Comment (GarrettWu) [reply bigframes/ml/llm.py:1038]: not really. We need df_fail for next retry, and append df_succ to df_result of current round. And I don't think DF subtract will be more effecient.

File: bigframes/ml/llm.py [hunk 3]
```diff
@@ -1018,15 +1029,37 @@ def predict(
...
+
+            if df_succ.empty:
+                warnings.warn("Can't make any progress, stop retrying.", RuntimeWarning)
+                break
+
+            df_result = (
+                bpd.concat([df_result, df_succ]) if not df_result.empty else df_succ
```
[the remark this answers is not in the document]
Comment (GarrettWu) [reply bigframes/ml/llm.py:1045]: It gave me error in tests, sth related to multi-index

File: bigframes/ml/llm.py [hunk 4]
```diff
@@ -1018,15 +1029,37 @@ def predict(
...
+        df_fail = X
+        for _ in range(max_retries + 1):
+            df = self._bqml_model.generate_text(df_fail, options)
 
-        if (df[_ML_GENERATE_TEXT_STATUS] != "").any():
+            df_succ = df[df[_ML_GENERATE_TEXT_STATUS].str.len() == 0]
+            df_fail = df[df[_ML_GENERATE_TEXT_STATUS].str.len() > 0]
+
+            if df_succ.empty:
+                warnings.warn("Can't make any progress, stop retrying.", RuntimeWarning)
```
[the remark this answers is not in the document]
Comment (GarrettWu) [reply bigframes/ml/llm.py:1041]: make sense.

**Pull request merged**